qiskit/synthesis/permutation/permutation_lnn.py

```diff
@@ -0,0 +1,68 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Depth-efficient synthesis algorithm for Permutation gates."""
+
+from qiskit import QuantumCircuit
+from .permutation_utils import _inverse_pattern
+
+
+def synth_permutation_depth_lnn_kms(pattern):
+    """
+    Permutation synthesis algorithm from [1], Chapter 6.
+
+    Synthesizes any permutation of n qubits over linear nearest-neighbor
+    architecture using SWAP gates with depth at most n and size at most
+    n(n-1)/2 (where both depth and size are measured with respect to SWAPs).
+
+    Args:
+        pattern (Union[list[int], np.ndarray]): permutation pattern.
+
+    Returns:
+        QuantumCircuit: the synthesized quantum circuit.
+
+    References:
+
+        [1]: Kutin, S., Moulton, D. P., Smithline, L. (2007).
+             Computation at a Distance.
+            `arXiv:quant-ph/0701194 <https://arxiv.org/abs/quant-ph/0701194>`_.
+    """
+
+    # In Qiskit, the permutation pattern [2, 4, 3, 0, 1] means that
+    # the permutation that maps qubit 2 to position 0, 4 to 1, 3 to 2, 0 to 3, and 1 to 4.
+    # In the permutation synthesis code below the notation is opposite:
+    # [2, 4, 3, 0, 1] means that 0 maps to 2, 1 to 3, 2 to 3, 3 to 0, and 4 to 1.
+    # This is why we invert the pattern.
+    cur_pattern = _inverse_pattern(pattern)
+
+    num_qubits = len(cur_pattern)
+    qc = QuantumCircuit(num_qubits)
+
+    # add conditional odd-even swap layers
+    for i in range(num_qubits):
+        _create_swap_layer(qc, cur_pattern, i % 2)
+
+    assert cur_pattern == sorted(cur_pattern)
```

Comment:
We should probably avoid bare asserts like this, they can be optimized away, but also there hard to debug for users because it just raises an `AssertionError` without any context. If you want this to be fatal we should us an if then raise, although I'm thinking we can get away with dropping this.

Reply:
Agreed, dropping the assert. And we have enough permutation synthesis tests to check correctness of this synthesis method.